index.ts

```diff
@@ -87,53 +84,110 @@ function getPackageVersion(pkg: PackageJson): string {
+interface ReadDirOpts {
+  linkDeps?: boolean;
+}
+
+// This only shallow-merges with any user-provided files, which is OK right now
+// because this is only one level deep. If we ever make it deeper, we'll need to
+// switch to a proper deep merge.
+const defaultFiles = {
+  'index.js': `
+    'use strict';
+     module.exports = {};`,
+};
+
+export interface ProjectArgs {
+  name?: string;
+  version?: string;
+  files?: fixturify.DirJSON;
 }
 
-class Project {
+export class Project {
   pkg: PackageJson;
-  files: fixturify.DirJSON = {
-    'index.js': `
-'use strict';
-module.exports = {};`,
-  };
+  files: fixturify.DirJSON;
   readonly isDependency = true;
 
   private _dependencies: { [name: string]: Project } = {};
   private _devDependencies: { [name: string]: Project } = {};
-  private _root: string;
+  private _baseDir: string | undefined;
   private _tmp: tmp.SynchrounousResult | undefined;
 
-  constructor(name: string, version = '0.0.0', cb?: (project: Project) => void, root?: string) {
-    this.pkg = {
-      name,
-      version,
-      keywords: [],
-    };
+  private dependencyLinks: Map<string, string> = new Map();
+  private linkIsDevDependency: Set<string> = new Set();
+
+  constructor(name?: string, version?: string, args?: Omit<ProjectArgs, 'name' | 'version'>);
+  constructor(name?: string, args?: Omit<ProjectArgs, 'name'>);
+  constructor(args?: ProjectArgs);
+  constructor(
+    first?: string | ProjectArgs,
+    second?: string | Omit<ProjectArgs, 'name'>,
+    third?: Omit<ProjectArgs, 'name' | 'version'>
+  ) {
+    let name: string | undefined;
+    let version: string | undefined;
+    let files: fixturify.DirJSON | undefined;
+
+    if (first == null) {
+      // all optional args stay undefined
+    } else if (typeof first === 'string') {
+      name = first;
+      if (typeof second === 'string') {
+        version = second;
+        if (third) {
+          ({ files } = third);
+        }
+      } else {
+        if (second) {
+          ({ version, files } = second);
+        }
+      }
+    } else {
+      ({ name, version, files } = first);
+    }
 
-    this.validate();
+    let pkg: PackageJson = {};
 
-    if (root) {
-      this._root = root;
-    } else {
-      this._tmp = tmp.dirSync({ unsafeCleanup: true });
-      this._root = fs.realpathSync(this._tmp.name);
+    if (files && typeof files?.['package.json'] === 'string') {
+      pkg = JSON.parse(files['package.json']);
+      files = Object.assign({}, files);
+      delete files['package.json'];
     }
 
-    if (typeof cb === 'function') {
-      cb(this);
+    this.pkg = Object.assign({}, pkg, {
+      name: name || pkg.name || 'a-fixturified-project',
+      version: version || pkg.version || '0.0.0',
+      keywords: pkg.keywords || [],
+    });
+    if (files) {
+      this.files = { ...defaultFiles, ...files };
+    } else {
+      this.files = defaultFiles;
     }
   }
 
-  get root() {
-    return this._root;
+  set baseDir(dir: string) {
```

Comment:
Consistency is good, but is this a breaking change? If so, would leaving a `get root() { }` accessor as an alias address that?

Reply:
This is a breaking and fundamental change.

A `Project` used to represent both your actual project directory *and that directory's parent*. That caused a lot of extra ceremony to deal with the package name (and any package scope) appearing as extra directories.

As of this PR, that's all gone. A `Project` represents only your actual project directory. So the thing that used to be `root` doesn't exist anymore.